keps/sig-storage/20190204-kubernetes-csi-release-process.md

```diff
@@ -0,0 +1,302 @@
+Kubernetes even though they are maintained by the Kubernetes project.
+
+This document explains how these components are released.
+
+## Motivation
+
+So far, the process for tagging and building components has been
+fairly manual, with some assistance by Travis CI. There has been no
+automatic testing on a Kubernetes cluster, neither stand-alone
+deployment of individual components nor full end-to-end (E2E) testing
+of several Kubernetes-CSI components.
+
+With CSI support reaching GA in Kubernetes 1.13, it is time to define
+and follow a less work-intensive and more predictable approach.
+
+### Goals
+
+- define the release process for each component
+- unit testing as mandatory pre-submit check for all components
+- testing on Kubernetes as mandatory pre-submit check for components that get
+  deployed in a cluster
+
+### Non-Goals
+
+- a combined "Kubernetes-CSI release": each component gets released
+  separately.  It is the responsibility of a CSI driver maintainer
+  pick and test sidecar releases for a combined deployment of that
+  driver. The [hostpath deployment
+  example](https://github.com/kubernetes-csi/csi-driver-host-path/tree/master/deploy)
+  can be used as a starting point, but it's not going to be able to
+  test all the possible combinations of features that a specific CSI
+  driver may need.
+- change responsibilities among the kubernetes-csi maintainers: as before, each
+  component will have some main maintainer who is responsible for releasing updates
+  of that component
+- automatically generate release notes: this will be covered by a separate, future KEP
+
+
+## Proposal
+
+### Versionioning
+
+Each of the components has its own documentation, versioning and
+release notes. [Semantic versioning](https://semver.org/) with `v`
+prefix (i.e. `v1.2.3`) is used for components that provide a stable
+API, like for example:
+- external-attacher
+- external-provisioner
+- external-snapshotter
+- cluster-driver-registrar
+- node-driver-registrar
+- csi-driver-host-path
+- csi-lib-utils
+
+Other components are internal utility packages that are
+not getting tagged:
+- csi-release-tools
+
+An additional suffix like `-rc1` can be used to denote a pre-release.
+
+The [hostpath example
+deployment](https://github.com/kubernetes-csi/csi-driver-host-path/tree/master/deploy)
+defines the exact release of each sidecar that is part of that
+deployment. When updating to newer sidecar releases, a new hostpath
+driver release is tagged with version numbers bumped according to the
+same rules as the individual components (major version bump when any
+of the sidecars had a major change, etc.).
+
+#### Release artifacts
+
+Tagging a component with a `v*` tag triggers a
+build for that component with that version. The output is primarily the container image
+for components that need to be deployed. Those images get published as
+`gcr.io/kubernetes-csi/<image>:<tag>`, which depends on [issue
```

Comment:
Lets make sure we build the binary for all the architectures like `amd, arm64, ppc64le`

Reply:
Can we leave that as a future enhancement?

Cross-compiling for other architectures and then publishing images wouldn't be so hard, but then we would also need some way of testing those. That's too much for the initial step.

Reply:
are cross builds being done now? (even if manual)
If so, is manual still going to be an option if put out of scope? should be explicitly stated.

Reply:
I'm not aware of anyone doing it.

Doing it manually is an option, but will require some work to adapt build rules.